pkg/roll/update_test.go

```diff
@@ -1108,6 +1132,93 @@ func TestRollLoopTypicalCase(t *testing.T) {
 	}
 }
 
+func TestRollLoopNilAuditLogDetails(t *testing.T) {
+	nodes := map[types.NodeName]bool{
+		"node1": true,
+		"node2": true,
+		"node3": true,
+	}
+	upd, _, manifest, _, f := updateWithHealth(t, 3, 0, nodes, nil, nil, nil, rc_fields.StaticStrategy)
+	defer f()
+	upd.DesiredReplicas = 3
+	upd.MinimumReplicas = 2
+
+	healths := make(chan map[types.NodeName]health.Result)
+	checks := map[types.NodeName]health.Result{
+		"node1": {Status: health.Passing},
+		"node2": {Status: health.Passing},
+		"node3": {Status: health.Passing},
+	}
+	upd.hcheck = cannedWatchServiceChecker{
+		watchServiceCh: healths,
+		serviceResult:  checks,
+	}
+	alerter := fakeAlerter{}
+	upd.alerter = alerter
+
+	ctx, _ := context.WithCancel(context.Background())
+
+	var wg sync.WaitGroup
+
+	rollLoopResult := make(chan bool)
+
+	// We replace upd.labeler with one that will return an error from GetLabels().
+	// That forces NewRUCompletionEventDetails to return nil details and an error
+	// so that we can test that the Roll returns early and does not call
+	// auditLogStore.Create() with nil details
+	fixture := consulutil.NewFixture(t)
+	upd.labeler = &badGetLabels{upd.ID(), *labels.NewConsulApplicator(fixture.Client, 0, 0)}
+
+	go func() {
+		rollLoopResult <- upd.Run(ctx)
+		close(rollLoopResult)
+	}()
+
+	healths <- checks
+
+	err := transferNode("node1", manifest, upd)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	healths <- checks
+
+	err = transferNode("node2", manifest, upd)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	healths <- checks
+
+	err = transferNode("node3", manifest, upd)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	healths <- checks
+
+	assertRollLoopResult(t, rollLoopResult, false)
+
+	wg.Wait()
+
+	// allow time for asynchronous upd.alerter.Alert() call, then
+	// confirm it was called
+	time.Sleep(1 * time.Second)
+	if alerter.numCalls != 0 {
```

Comment:
oh is this supposed to be 1? basing that on your comments

Reply:
Yes this is supposed to 1. That means the test is broken, possibly due to what you said above. Great catch!